File: src/csr_preconilu0.f
!                   *************************
                    SUBROUTINE CSR_PRECONILU0
!                   *************************
     & (LUV,MA)
!
!***********************************************************************
! CSC PACKAGE - SERGIO CASTIBLANCO
!***********************************************************************
!
!brief    1) COMPUTES THE INCOMPLETE-LU(0) DECOMPOSITION OF MA AND
!            RETURN THE VALUES IN LUV, CSR STORAGE
!
!history  Sergio Castiblanco
!+        21/02/2022
!+        Translation for original Matlab implementation
!
!+
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!| LUV      |<->| VECTOR OF VALUES OF LU MATRIX IN CSC                 |
!| MA       |-->| MATRIX A IN CSR                                      |
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      USE CSC_STORAGE, EX_CSR_PRECONILU0 => CSR_PRECONILU0
      IMPLICIT NONE
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      TYPE(CSC_OBJ), INTENT(IN) :: MA
      DOUBLE PRECISION, DIMENSION(MA%NZ), INTENT(INOUT) :: LUV
!
!  IN SUBROUTINE VARIABLES
!
      DOUBLE PRECISION, DIMENSION(MA%NZ) :: AV
      INTEGER, DIMENSION(MA%NZ) :: AC
      INTEGER, DIMENSION(MA%NC+1) :: AR
      INTEGER, DIMENSION(MA%NC) :: DIA, POINT
      INTEGER :: N,I,J,K,W,V
!
! COMPUTING ILU0 FACTORIZATION (MITTAL & AL-KURDI, 2003)
!
      AV = MA%V
      AR = MA%C
      AC = MA%R
      LUV = AV
!
      N = SIZE(AR)-1
      DIA = 0
      DO I=1,N
        DO J=AR(I),(AR(I+1)-AR(1))
          IF(AC(J).EQ.I)THEN
            DIA(I)=J
          ENDIF
        ENDDO
      ENDDO
      POINT = 0
!
      DO I=2,N
        DO V=AR(I)+1,(AR(I+1)-AR(1))
          POINT(AC(V))=V
        ENDDO
        DO V=AR(I),DIA(I)-1
          J=AC(V)
          LUV(V) = LUV(V)/LUV(DIA(J))
          DO W = (DIA(J)+1),(AR(J+1)-AR(1))
            K = POINT(AC(W))
            IF(K.GT.0)THEN
              LUV(K) = LUV(K) - LUV(V)*LUV(W)
            !ELSE   !IF REVISED, IF NOT (DIAGONAL MAY HAVE ZEROS)
            !  LUV(DIA(I)) = LUV(DIA(I)) - LUV(V)*LUV(W)
            ENDIF
          ENDDO
        ENDDO
        DO V=AR(I)+1,(AR(I+1)-AR(1))
          POINT(AC(V))=0
        ENDDO
      ENDDO
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      END SUBROUTINE CSR_PRECONILU0

File: matlab/csr_preconILU0.m
function [LUv] = csr_preconILU0(Av,Ac,Ar)
%
% This function takes the matrix A in CSC storage, and prepares the
% matrices for solve the system Ax=b with ILU0 preconditioner
%
% L and U:
% The ILU0 preconditioner is defined with the next matrix:
% 
%               Milu0 = L*U
%
% Where L and U are the result of an Incomplete LU factorization, with the
% same non-zero structure as matrix A. Since L is a unitary lower
% triangular matrix. The ILU0 can be stored in packed format, and, as it
% has the same structure as A, it is only needed to compute the vector of
% values LUv. Additionaly
%
% In which the system can be solved in the next way:
%
%               solve Ly = x;  -> Forward substitution
%               solve Uz = y;  -> Backward substitution
%
% Entries:
%     Av, Ar, Ac : Components of matrix A in CSC stotage
%
%      Sergio A. Castiblanco B. - Métodos Numéricos Avanzados
% Original algorithm: (Mittal & Al-Kurdi, 2003)
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

LUv = Av;

%Incomplete LU sparse CSR (Mittal Al-Kurdi, 2003)
n = length(Ar)-1;
dia = zeros(n,1);
for i=1:n
    for j=Ar(i):Ar(i+1)-1
        if Ac(j) == i
            dia(i) = j;
        end
    end
end
point = zeros(n,1);

for i = 2:n
    for v = Ar(i)+1:Ar(i+1)-1
        point(Ac(v))=v;
    end
    for v = Ar(i):dia(i)-1
        j = Ac(v);
        LUv(v) = LUv(v)/LUv(dia(j));    
        for w = dia(j)+1:Ar(j+1)-1
            k = point(Ac(w));
            if k>0
                LUv(k) = LUv(k) - LUv(v)*LUv(w);
%             end
            %else
            %   LUv(dia(i)) = LUv(dia(i)) - LUv(v)*LUv(w);
            end
        end
    end
    for v=Ar(i)+1:Ar(i+1)-1
        point(Ac(v))=0;
    end
end

end